### PR TITLE
[3581] Trainees with invalid TRNs

### PR DESCRIPTION
**Pull request by HettieS**
### Context

It was picked up during the DTTP imports that there were some additional invalid TRNs coming from DTTP that needed to be set as nil in Register. Arlston found these with a regexp and I've implemented that in the code.

We have discussed creating a new metadata column to store the invalid TRNs as these are sometimes used by Ops team as an indication of why a trainee is missing a TRN e.g. `999999999` indicates the trainee came from HESA without a TRN.

Research for which TRNs to include is linked to this trello card: https://trello.com/c/6wuGuyu6/3579-check-do-all-post-submission-trainees-have-trns

### Changes proposed in this pull request

* Add TRN_REGEX rule so that we only accept TRNs that are 6 or 7 digits
* Invalid TRNs are saved as nil
* Remove INVALID_TRN constant as this is now covered by the regex rule
* Trim whitespace in `trn` method in `models/dttp/trainee.rb`
* Add `:with_invalid_trn` as a trait to the `api_trainee` factory

### Guidance to review

* Look at the implementation of the regex and see if you think I've missed any edge cases! 
* Test DTTP import on this?

### Important business

* ~~Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?~~
* ~~Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?~~

### CODE REVIEW

**Comment by HettieS** [comment]
[Trainees with invalid TRNs](https://trello.com/c/e49x8x2j/3581-trainees-with-invalid-trns)

File: app/services/trainees/create_from_dttp.rb [hunk 1]
```diff
@@ -6,6 +6,9 @@ class CreateFromDttp
     include HasDttpMapping
 
     INVALID_TRN = "999999999"
+
+    TRN_REGEX = /^(\s*[0-9]+[.]?[0-9]*\s*|\s*[.][0-9]+\s*)$/.freeze
```
Comment (AbigailMcP) [inline app/services/trainees/create_from_dttp.rb:10]: Instead of adding the whitespace to the regex rule, how about matching a stripped version of `dttp_trainee.trn` below on line 143?

By adding it to our regex rule it feels like we're saying whitespace is 'correct'... which strictly it's not.

The outcome might be the same but I feel like keeping the regex strictly correct is nicer. You'll see a lot of codebases with `something.strip` before continuing with other validations.

Comment (HettieS) [reply app/services/trainees/create_from_dttp.rb]: makes sense! I've now modified the regex rule and how I've implemented `strip` , let me know what you think!

Also chatted to Arlston and we agreed the regex can be simpler (we didn't need to account for decimal point for example)

File: app/services/trainees/create_from_dttp.rb [hunk 2]
```diff
@@ -135,8 +138,19 @@ def undergrad_level?
...
 
+    def trn_present?
+      dttp_trainee.trn.present?
+    end
+
     def trn
-      dttp_trainee.trn == INVALID_TRN ? nil : dttp_trainee.trn
+      if trn_present?
+
+        dttp_trainee.trn.strip!
```
Comment (archferns) [inline app/services/trainees/create_from_dttp.rb:148]: using `strip!` might modify the original trn value in the database against `response->'dfe_trn'` if whitespace exists.
would it be possible to try this with the non-bang version please?

File: app/services/trainees/create_from_dttp.rb [hunk 3]
```diff
@@ -6,6 +6,9 @@ class CreateFromDttp
     include HasDttpMapping
 
     INVALID_TRN = "999999999"
+
+    TRN_REGEX = /^([0-9]{7,})$/.freeze
```
Comment (archferns) [inline app/services/trainees/create_from_dttp.rb:10]:
```suggestion
    TRN_REGEX = /^\d{7}$/.freeze
```
https://www.gov.uk/guidance/teacher-reference-number-trn seems to suggest that a TRN is always a 7 digit number.
if that is definitely what we expect, this should also take care of the `999999999`, and we wouldn't need a special handling for it.

Comment (HettieS) [reply app/services/trainees/create_from_dttp.rb]: thanks Arlston - after doing some research on this, looking at existing `dfe_trn` variations from DTTP and speaking to Rob Collier who works with DQT, we decided to accept TRNs that are 6 or 7 digits long - historically, it is possible TRNs with less than 7 digits might have dropped leading/trailing 0s. So we'll save those ones too in case they are valid.

Comment (archferns) [reply app/services/trainees/create_from_dttp.rb]: good idea, makes sense about historical values

**Review by archferns — APPROVED** [review]
Looks great!
